README.md

```diff
@@ -2,4 +2,196 @@
+
+```typescript
+const validationResult = validateObjectShape(
+    objectDescription,
+    validationItem,
+    expectedObjectShape,
+    validationOptions
+)
+
+// If the validation is successful
+{
+    valid: true,
+    result: { ...validatedItem }  // (the object you passed in, but typed!)
+}
+
+// If the validation fails
+{
+    valid: false,
+    errors: [
+        'Expected Test obj.two to be type: number, was string',
+    ]
+}
+```
+
+`validateObjectShape` will coerce values for you if it's able to. For example if you say the property `age` is a number, but it's passed in as a string, `validateObjectShape` will try convert it to a number for you in the `result` object:
+
+```typescript
+const validationResult = validateObjectShape(
+    'Coercion Example',
+    { age: '30' },
+    { age: number }
+)
+
+if (validationResult.valid === true) {
+    // validationResult.result.age -> 30
+    typeof validationResult.result.age === 'number'
+}
+```
+
+The library also alows you to test nested objects and arrays, for example:
+
+```typescript
+const validationResult = validateObjectShape(
+    'Nested Example',
+    {
+        age: '30',
+        meta: { group: 'Staff' }
+        names: [
+            { fname: 'Jeff', lname: 'Thompson' },
+            { fname: 'Jeff', lname: 'Thompson' }
+        ]
+    },
+    {
+        age: number,
+        meta: { group: 'string' }
+        names: arrayOf({ fname: 'string', lname: 'string' })
+    }
+)
+```
+
+### Validation Types
+
+There are a number of basic validation types available to use, and some helpers which allow you build more complex types:
+
+Basic types
+
+-   **string**: Matches a string
+-   **number**: Matches a number
+-   **boolean**: Matches a boolean
+-   **unknown**: Matches an unknown type (will skip validation on that property)
+
+This package also understands more complex types such as arrays and optional types. You can import helper functions to assist you in building these.
+
+Complex Types:
+
+-   **arrayOf** ('string'): Array of some basic type
+-   **optional** ('string'): Makes the property your testing optional. The test will only run if the property exists
+
+```typescript
+const result = validateObjectShape(
+    'Test obj',
+    {
+        one: 'string value',
+        two: 2,
+        three: true,
+        four: ['1', '2'],
+        five: [true, false],
+        six: [1, 2],
+        seven: 'whatever'
+    },
+    {
+        one: 'string',
+        two: 'number',
+        three: 'boolean',
+        four: arrayOf('string'),
+        five: arrayOf('boolean'),
+        six: arrayOf('number'),
+        seven: 'unknown',
+        eight: optional('boolean')
+    }
+)
+```
+
+### Options
+
+-   **coerceValidObjectIntoArray** (boolean): If you are validating an array with `arrayOf`, setting this to `true` will convert those properties to arrays for you, rather than returning an error.
```

Comment:
Could always put a comment that this is useful when using xml to json libraries like ....

Reply:
Good idea, added a note.